File: UtilsMethod.js
const prompt = require("prompt-sync")();
class UtilsMethod {
  static askUser(messages) {
    messages.forEach((message) => console.log(message));
    const input = prompt("");
    return this.takeIntValueFromString(input);
  }

  static askDate() {
    console.log("Année ");
    let year = this.takeIntValueFromString(prompt(""));
    while (year < 0) {
      console.log("Année invalide, réécrire l'année");
      year = this.takeIntValueFromString(prompt(""));
    }
    console.log("Mois");
    let month = this.takeIntValueFromString(prompt(""));
    while (month < 0 || month > 12) {
      console.log("Mois invalide, réécrire le mois");
      month = this.takeIntValueFromString(prompt(""));
    }
    console.log("Jour");
    let day = this.takeIntValueFromString(prompt(""));
    while (day < 0 || day > 28) {
      console.log("Jour invalide, réécrire le jour");
      day = this.takeIntValueFromString(prompt(""));
    }
    return new Date(year, month - 1, day);
  }

  static useUniqueItem(uniqueItem, actualFunction, createUniqueItemFunction) {
    if (uniqueItem === null) {
      const choice = this.askUser([
        "1. Ajouter un item",
        "2. Revenir au menu principal",
      ]);

      switch (choice) {
        case 1:
          createUniqueItemFunction("actualWallet");
          break;
        case 2:
          break;
        default:
          console.log("Choix invalide. Veuillez réessayer.");
          actualFunction();
          break;
      }
    } else {
      const choice = this.askUser([
        "Quel événement s'est passé?",
        "1. Utilisé",
        "2. Revenir au menu principal",
      ]);

      switch (choice) {
        case 1:
          uniqueItem.usedItem(uniqueItem.toString());
          break;
        case 2:
          break;
        default:
          console.log("Choix invalide. Veuillez réessayer.");
          actualFunction();
          break;
      }
    }
    return null;
  }

  static useMultiItems(
    multipleItem,
    actualFunction,
    createMultiItemFunction,
    useFunction,
    addFunction,
  ) {
    if (multipleItem === null) {
      const choice = this.askUser([
        "1. Ajouter un item",
        "2. Revenir au menu principal",
      ]);

      switch (choice) {
        case 1:
          createMultiItemFunction("actualWallet");
          break;
        case 2:
          break;
        default:
          console.log("Choix invalide. Veuillez réessayer.");
          actualFunction();
          break;
      }
    } else {
      const choice = this.askUser([
        "1. Utiliser",
        "2. Ajouter",
        "3. Consulter la quantité actuelle",
        "4. Revenir au menu principal",
      ]);

      switch (choice) {
        case 1:
          console.log("Combien voulez-vous utiliser?");
          const quantityUsed = this.takeIntValueFromString(prompt());
          if (multipleItem.getQuantity() < quantityUsed) {
            console.log("Vous n'avez pas la quantité suffisante");
          } else {
            useFunction(quantityUsed);
            console.log("Opération réussie");
            console.log("Votre compte est de");
            console.log(
              `${multipleItem.getQuantity()} ${multipleItem.getUnity()}`,
            );
          }
          break;
        case 2:
          console.log("Combien voulez-vous ajouter?");
          const quantityToAdd = this.takeIntValueFromString(prompt());
          addFunction(quantityToAdd);
          console.log("Opération réussie");
          console.log("Votre compte est de");
          console.log(
            `${multipleItem.getQuantity()} ${multipleItem.getUnity()}`,
          );
          break;
        case 3:
          console.log("Votre compte est de");
          console.log(multipleItem.getQuantityWithUnity());
          break;
        case 4:
          break;
        default:
          console.log("Choix invalide. Veuillez réessayer.");
          actualFunction();
          break;
      }
    }
    return null;
  }

  static isNotConvertibleToNumber(str) {
    try {
      const number = parseInt(str);
      return isNaN(number) || number < 0;
    } catch (e) {
      return true;
    }
  }

  static takeIntValueFromString(str) {
    let value = str;
    let isValid;
    do {
      if (this.isNotConvertibleToNumber(value)) {
        isValid = false;
      } else {
        isValid = parseInt(value) >= 0;
      }
      if (!isValid) {
        console.log("Entrée invalide, veuillez entrer une nouvelle valeur.");
        value = prompt("");
      }
    } while (!isValid);
    return parseInt(value);
  }
}

module.exports = UtilsMethod;

File: UtilsMethod.py
from datetime import date
from typing import List, Callable


class UtilsMethod:
    @staticmethod
    def ask_user(messages: List[str]) -> int:
        for message in messages:
            print(message)
        return UtilsMethod.take_int_value_from_string(input())

    def ask_date(self) -> date:
        print("Année")
        year = UtilsMethod.take_int_value_from_string(input())
        while year < 0:
            print("Année invalide, réécrire l'année")
            year = UtilsMethod.take_int_value_from_string(input())

        print("Mois")
        month = UtilsMethod.take_int_value_from_string(input())
        while month < 0 or month > 12:
            print("Mois invalide, réécrire le mois")
            month = UtilsMethod.take_int_value_from_string(input())

        print("Jour")
        day = UtilsMethod.take_int_value_from_string(input())
        while day < 0 or day > 28:
            print("Jour invalide, réécrire le jour")
            day = UtilsMethod.take_int_value_from_string(input())

        return date(year, month, day)

    @staticmethod
    def use_unique_item(unique_item, actual_function: Callable, create_unique_item_function: Callable):
        if unique_item is None:
            choice = UtilsMethod.ask_user(["1. Ajouter un item", "2. Revenir au menu principal"])

            if choice == 1:
                create_unique_item_function("actualWallet")
            elif choice == 2:
                print("revenir")
            else:
                actual_function()
        else:
            choice = UtilsMethod.ask_user(["Quel événement s'est passé?", "1. Utilisé", "2. Revenir au menu principal"])

            if choice == 1:
                unique_item.used_item(unique_item.__str__())
            elif choice == 2:
                print("revenir")
            else:
                actual_function()

    def use_multi_items(
        multiple_item,
        actual_function: Callable,
        create_multi_item_function: Callable,
        use_function: Callable,
        add_function: Callable,
    ):
        if multiple_item is None:
            choice = UtilsMethod.ask_user(["1. Ajouter un item", "2. Revenir au menu principal"])

            if choice == 1:
                create_multi_item_function("actualWallet")

        else:
            choice = UtilsMethod.ask_user(
                ["1. Utiliser", "2. Ajouter", "3. Consulter la quantité actuelle", "4. Revenir au menu principal"]
            )

            if choice == 1:
                print("Combien voulez-vous utiliser?")
                quantity_used = UtilsMethod.take_int_value_from_string(input())
                if multiple_item.get_quantity() < quantity_used:
                    print("Vous n'avez pas la quantité suffisante")
                else:
                    use_function(quantity_used)
                    print("Opération réussie")
                    print("Votre compte est de")
                    print(f"{multiple_item.get_quantity() - quantity_used} {multiple_item.get_unity()}")

            elif choice == 2:
                print("Combien voulez-vous ajouter?")
                quantity_to_add = UtilsMethod.take_int_value_from_string(input())
                add_function(quantity_to_add)
                print("Opération réussie")
                print("Votre compte est de")
                print(f"{multiple_item.get_quantity() + quantity_to_add} {multiple_item.get_unity()}")

            elif choice == 3:
                print("Votre compte est de")
                print(multiple_item.get_quantity_with_unity())

    @staticmethod
    def is_not_convertible_to_number(string_value: str) -> bool:
        try:
            number = int(string_value)
            return number <= 0
        except ValueError:
            return True

    @staticmethod
    def take_int_value_from_string(string_value: str) -> int:
        value = string_value
        is_valid = not UtilsMethod.is_not_convertible_to_number(value)
        while not is_valid:
            print("Entrée invalide, veuillez entrer une nouvelle valeur.")
            value = input()
            is_valid = not UtilsMethod.is_not_convertible_to_number(value) and int(value) >= 0
        return int(value)

    def __init__(self):
        pass
